File: packages/contracts/src/usage.ts
import { z } from "zod";

import { nullableField, SCHEMA_VERSION, schemaVersionField } from "./common.js";

/**
 * "A song was displayed during a service." The canonical cross-app usage event,
 * emitted by SundayStage (or Plan) and recorded by SundaySong as the source of
 * truth for CCLI + TONO reporting. Mirrors SundaySong's `UsageLogInputSchema`
 * (`/v1/usage/log`); the API dedupes on `idempotency_key`.
 *
 * `was_streamed` is the critical bit: streamed performances feed a different
 * royalty pool than in-room ones.
 */
export const UsageEvent = z.object({
  schema_version: schemaVersionField,
  church_id: z.string().uuid(),
  song_id: z.string().uuid(),
  variant_id: nullableField(z.string().uuid()),
  /** ISO calendar date YYYY-MM-DD. */
  service_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
  duration_displayed_sec: nullableField(z.number().int().min(0)),
  was_streamed: z.boolean(),
  idempotency_key: z.string().min(8).max(120),
});
export type UsageEvent = z.infer<typeof UsageEvent>;

/**
 * Build a deterministic idempotency key for a usage event so a re-sent event
 * (network retry, app restart) never double-counts. Stable for a given service
 * item.
 */
export function makeUsageIdempotencyKey(serviceId: string, serviceItemId: string): string {
  return `svc-${serviceId}:item-${serviceItemId}`;
}

/** Inputs for {@link buildUsageEvent} — the Stage→Song usage bridge. */
export interface BuildUsageEventInput {
  churchId: string;
  songId: string;
  variantId?: string | null;
  /** ISO calendar date YYYY-MM-DD. */
  serviceDate: string;
  wasStreamed: boolean;
  durationDisplayedSec?: number | null;
  /** The service this song was shown in — feeds the idempotency key. */
  serviceId: string;
  /** The running-order item — feeds the idempotency key. */
  serviceItemId: string;
}

/**
 * Build a validated {@link UsageEvent} from a service item, deriving the dedupe
 * key with {@link makeUsageIdempotencyKey} so a retried emit never double-counts.
 * The canonical way SundayStage/Plan report a played song to SundaySong's
 * `/v1/usage/log`.
 */
export function buildUsageEvent(input: BuildUsageEventInput): UsageEvent {
  return UsageEvent.parse({
    schema_version: SCHEMA_VERSION,
    church_id: input.churchId,
    song_id: input.songId,
    variant_id: input.variantId ?? null,
    service_date: input.serviceDate,
    duration_displayed_sec: input.durationDisplayedSec ?? null,
    was_streamed: input.wasStreamed,
    idempotency_key: makeUsageIdempotencyKey(input.serviceId, input.serviceItemId),
  });
}

File: crates/sunday-contracts/src/usage.rs
use serde::{Deserialize, Serialize};

use crate::common::{default_schema_version, SCHEMA_VERSION};

/// "A song was displayed during a service." Mirrors SundaySong's
/// `UsageLogInputSchema` (`/v1/usage/log`); the API dedupes on `idempotency_key`.
/// `was_streamed` selects the royalty pool (streamed vs in-room).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct UsageEvent {
    #[serde(default = "default_schema_version")]
    pub schema_version: u32,
    pub church_id: String,
    pub song_id: String,
    pub variant_id: Option<String>,
    /// ISO calendar date YYYY-MM-DD.
    pub service_date: String,
    pub duration_displayed_sec: Option<i64>,
    pub was_streamed: bool,
    pub idempotency_key: String,
}

/// Deterministic idempotency key so a re-sent usage event never double-counts.
pub fn make_usage_idempotency_key(service_id: &str, service_item_id: &str) -> String {
    format!("svc-{service_id}:item-{service_item_id}")
}

/// Inputs for [`build_usage_event`] — the Stage→Song usage bridge. Mirrors the
/// TypeScript `BuildUsageEventInput`.
pub struct BuildUsageEventInput<'a> {
    pub church_id: &'a str,
    pub song_id: &'a str,
    pub variant_id: Option<&'a str>,
    /// ISO calendar date YYYY-MM-DD.
    pub service_date: &'a str,
    pub was_streamed: bool,
    pub duration_displayed_sec: Option<i64>,
    /// The service this song was shown in — feeds the idempotency key.
    pub service_id: &'a str,
    /// The running-order item — feeds the idempotency key.
    pub service_item_id: &'a str,
}

/// Build a [`UsageEvent`] from a service item, deriving the dedupe key with
/// [`make_usage_idempotency_key`]. Mirrors the TypeScript `buildUsageEvent`.
pub fn build_usage_event(input: &BuildUsageEventInput) -> UsageEvent {
    UsageEvent {
        schema_version: SCHEMA_VERSION,
        church_id: input.church_id.to_string(),
        song_id: input.song_id.to_string(),
        variant_id: input.variant_id.map(str::to_string),
        service_date: input.service_date.to_string(),
        duration_displayed_sec: input.duration_displayed_sec,
        was_streamed: input.was_streamed,
        idempotency_key: make_usage_idempotency_key(input.service_id, input.service_item_id),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_usage_event_derives_stable_key() {
        let input = BuildUsageEventInput {
            church_id: "c1",
            song_id: "s1",
            variant_id: None,
            service_date: "2026-05-31",
            was_streamed: true,
            duration_displayed_sec: None,
            service_id: "svc-9",
            service_item_id: "1",
        };
        let e = build_usage_event(&input);
        assert_eq!(e.idempotency_key, make_usage_idempotency_key("svc-9", "1"));
        assert_eq!(e.schema_version, SCHEMA_VERSION);
        assert!(e.variant_id.is_none());
        assert!(e.was_streamed);
    }

    #[test]
    fn build_usage_event_carries_optional_fields() {
        let input = BuildUsageEventInput {
            church_id: "c1",
            song_id: "s1",
            variant_id: Some("v1"),
            service_date: "2026-05-31",
            was_streamed: false,
            duration_displayed_sec: Some(312),
            service_id: "svc-9",
            service_item_id: "2",
        };
        let e = build_usage_event(&input);
        assert_eq!(e.variant_id.as_deref(), Some("v1"));
        assert_eq!(e.duration_displayed_sec, Some(312));
    }
}
